# Sur les nombres à virgule \[flottante\] ...

La plupart des languages de programmation ont deux systèmes de nombres : 
* un système pour les nombres entiers (représentés en base 2) et 
* un système pour les nombres à virgule (représentés selon la norme IEEE 754)

Dans cette note, je regarderais les nombres à virgule \[flottante\]. 

C'est une expérience similaire à faire 1/3 + 1/3 + 1/3 en base 10 : 0.333... + 0.333.... + 0.333... = 0.999... que l'on peut faire avec ces nombres :

In [4]:
0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1

0.9999999999999999

Si l'on regarde de plus près, l'erreur se produit pour 0.1, mais aussi pour d'autres nombres :

In [6]:
l=[0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9]
for e in l:
    s = 0
    for i in range(10):
        s = s + e
    print(e, "+", e , "+ ... + ", e, "=", s)

0.1 + 0.1 + ... +  0.1 = 0.9999999999999999
0.2 + 0.2 + ... +  0.2 = 1.9999999999999998
0.3 + 0.3 + ... +  0.3 = 2.9999999999999996
0.4 + 0.4 + ... +  0.4 = 3.9999999999999996
0.5 + 0.5 + ... +  0.5 = 5.0
0.6 + 0.6 + ... +  0.6 = 5.999999999999999
0.7 + 0.7 + ... +  0.7 = 7.000000000000001
0.8 + 0.8 + ... +  0.8 = 7.999999999999999
0.9 + 0.9 + ... +  0.9 = 9.000000000000002


Finalement, le plus étrange, c'est pour 0.5, le calcul est juste. Est-ce par hasard ? 

En fait, non. Tout comme 1/3 a une écriture infinie en base 10, mais 0.1 a une écriture finie en base 10 ; le nombre 0.1 a une écriture infinie en base 2 (si l'on prends les nombres après la virgule sur lesquels se fonde la norme IEEE 754) mais contre 0.5 a une écriture finie en  base 2 : 0.5 c'est $2^{-1}$, donc en base 2 cela s'ecrit 0.1|$_{base_2}$.

Revenons à 0.1 et aux autres nombres ...

Il y a un autre calcul que l'on peut faire pour voir à quel point le calcul peut être juste ou pas :

In [32]:
s = 0.1
t = "0.1"
for i in range(9):
    s = s + 0.1
    t = t + " + 0.1"
    print(t, " = ", s)

0.1 + 0.1  =  0.2
0.1 + 0.1 + 0.1  =  0.30000000000000004
0.1 + 0.1 + 0.1 + 0.1  =  0.4
0.1 + 0.1 + 0.1 + 0.1 + 0.1  =  0.5
0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1  =  0.6
0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1  =  0.7
0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1  =  0.7999999999999999
0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1  =  0.8999999999999999
0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1 + 0.1  =  0.9999999999999999


Ici, les calculs justes sont un heureux hasard ... (et si on continuait, cet heureux hasard arriverait de moins en moins souvent)

In [57]:
s = 0
for i in range(1,1000000):
    s = s + 0.1
    if len(" "+str(s))<10:
        print("0.1 + 0.1 + ... + 0.1 = ", s)

0.1 + 0.1 + ... + 0.1 =  0.1
0.1 + 0.1 + ... + 0.1 =  0.2
0.1 + 0.1 + ... + 0.1 =  0.4
0.1 + 0.1 + ... + 0.1 =  0.5
0.1 + 0.1 + ... + 0.1 =  0.6
0.1 + 0.1 + ... + 0.1 =  0.7
0.1 + 0.1 + ... + 0.1 =  1.2
0.1 + 0.1 + ... + 0.1 =  1.3
0.1 + 0.1 + ... + 0.1 =  4.4
0.1 + 0.1 + ... + 0.1 =  4.5
0.1 + 0.1 + ... + 0.1 =  4.6
0.1 + 0.1 + ... + 0.1 =  18.9
0.1 + 0.1 + ... + 0.1 =  19.0
0.1 + 0.1 + ... + 0.1 =  19.1
0.1 + 0.1 + ... + 0.1 =  75.2
0.1 + 0.1 + ... + 0.1 =  75.3
0.1 + 0.1 + ... + 0.1 =  301.2
0.1 + 0.1 + ... + 0.1 =  301.3
0.1 + 0.1 + ... + 0.1 =  1204.4
0.1 + 0.1 + ... + 0.1 =  1204.5
0.1 + 0.1 + ... + 0.1 =  1204.6
0.1 + 0.1 + ... + 0.1 =  4818.9
0.1 + 0.1 + ... + 0.1 =  4819.0
0.1 + 0.1 + ... + 0.1 =  4819.1
0.1 + 0.1 + ... + 0.1 =  19275.2
0.1 + 0.1 + ... + 0.1 =  19275.3
0.1 + 0.1 + ... + 0.1 =  77101.2
0.1 + 0.1 + ... + 0.1 =  77101.3


Donc 0.1+0.1+0.1 donne aussi une erreur (petite, mais une erreur tout de même). Et alors ?  peut-on partir sur cette erreur pour faire pire dans l'addition 0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1 ? Essayons :

In [20]:
(0.1+0.1+0.1)+(0.1+0.1+0.1)+(0.1+0.1+0.1)+0.1

1.0000000000000002

C'est pire si l'on regarde le détail :

In [25]:
premier_un = 0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1
print("En premier, erreur de :",premier_un-1)
second_un = (0.1+0.1+0.1)+(0.1+0.1+0.1)+(0.1+0.1+0.1)+0.1
print("En second, erreur de :",second_un-1)

En premier, erreur de : -1.1102230246251565e-16
En second, erreur de : 2.220446049250313e-16


L'erreur a doublé.

De plus, c'est étonnant l'erreur a changé de coté, précédement on arrivait à 0.9999999999999999 (moins que 1) alors que maintenant on arrive à 1.0000000000000002 (plus que 1).

## et alors ?

Et alors, donc dans tous les cas précédents, si l'on fait un test d'égalité, on obtient toujours le même résultat :

In [62]:
print("(0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1)==1 : ",(0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1)==1)
print("((0.1+0.1+0.1)+(0.1+0.1+0.1)+(0.1+0.1+0.1)+0.1)==1 : ",((0.1+0.1+0.1)+(0.1+0.1+0.1)+(0.1+0.1+0.1)+0.1)==1)
print("(0.1 + 0.1 + 0.1)==0.3 : ",(0.1 + 0.1 + 0.1)==0.3)


(0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1+0.1)==1 :  False
((0.1+0.1+0.1)+(0.1+0.1+0.1)+(0.1+0.1+0.1)+0.1)==1 :  False
(0.1 + 0.1 + 0.1)==0.3 :  False


## Conclusion

Attention aux nombres à virgule \[flottante\], en particulier lors de tests d'égalité.

ps : une autre fois, on pourra regarder :

In [80]:
x=10000000000000000
(x+0.1)-x

0.0